---
title: "Create a Monotonic Genetic Map"
author: "Brian Ward"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Create a Monotonic Genetic Map}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

The 'bwardr' package includes two genetic map files for the International Wheat Genome Sequencing Consortium's (IWGSC) v1.0 reference wheat genome. These genetic maps were created by the labs of Jesse Poland and Gary Muehlbauer, by aligning SNPs generated from the Synthetic x Opata cross using a *PstI-MspI* genotyping-by-sequencing (GBS) protocol against the reference genome. Note that these maps include the raw physical and genetic positions. This may cause problems in some applications, because the maps are not monotonic (i.e. genetic positions do not always uniformly increase with physical position). One way to address this is to fit a monotonic polynomial regression to the data for each chromosome, and then replace the raw genetic positions with the fitted values.

## Packages and Data

We will require the synop_dh88 data from package 'bwardr', as well as the package 'MonoPoly' to perform the monotonic regression.
```{r packages and data}
library(bwardr)
library(MonoPoly)
data("synop_dh88")
str(synop_dh88)
```

## Raw data plot

Below is a plot of the raw genetic map for chromosome 1A.
```{r plot raw data}
chr1a <- synop_dh88[synop_dh88$chr == "1A", ]
plot(chr1a$dist ~ chr1a$pos, 
     xlab = "Physical Position (bp)", 
     ylab = "Genetic Position (cM)")
```

Note that there are several "knots" in the plot, where the genetic distance decreases from one marker to the next.

## Monotonic regression

We can split the map by chromosome, and then run the monotonic regression for each chromosome. The degree argument of the monpol() function controls the degree of the fit polynomial. Here we use a degree of 9, which appears to suitably capture the curvature. The individual chromosomes are recombined, and the "dist" column (genetic distance) is replaced by the fitted values.
```{r chrom regressions}
maplist <- split(synop_dh88, synop_dh88$chr)
for (i in names(maplist)) {
  monreg <- monpol(dist ~ pos, data = maplist[[i]], degree = 9)
  maplist[[i]]$fitted <- as.vector(predict(monreg, newdata = maplist[[i]], scale = "original"))
}

mono_map <- do.call("rbind", maplist)
mono_map$dist <- round(mono_map$fitted, 2)
mono_map$fitted <- NULL
```

## Fitted values plot

Below is a plot of the physical position vs. the fitted genetic position values for chromosome 1A:
```{r plot fitted vals}
chr1a <- mono_map[mono_map$chr == "1A", ]
plot(chr1a$dist ~ chr1a$pos, 
     xlab = "Physical Position (bp)", 
     ylab = "Genetic Position (cM)")
```

Note that the genetic position of the fitted data now uniformly increases as the physical position increases.